lang/pkl/Portfile

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem              1.0
+PortGroup               java 1.0
+PortGroup               github 1.0
+
+github.setup            apple pkl 0.25.1
+
+java.version            17+
+java.fallback           openjdk17
+
+categories              lang
+maintainers             {sub-pop.net:link @subpop} openmaintainer
+
+license                 Apache-2.0
+
+description             A configuration as code language with rich validation and tooling
+long_description        Pkl (pronounced Pickle) is an embeddable configuration \
+                        language which provides rich support for data templating \
+                        and validation. It can be used from the command line, \
+                        integrated in a build pipeline, or embedded in a program. \
+                        Pkl scales from small to large, simple to complex, ad-hoc \
+                        to repetitive configuration tasks.
+
+checksums               rmd160  1e822e7a57050b392a216d9a808283d42df3801b \
+                        sha256  f1588550d820e226b9c387abcffae50c4c64ee313a872acdefbe2f61b6f60e84 \
+                        size    2769668
+
+if {${os.arch} eq "arm"} {
+patchfiles              graalVm23.patch
+patch.args              -p1
+}
+
+use_configure           no
+
+supported_archs         arm64 x86_64
```

Comment:
this doesn't support `i386` or any other possible architecture? If not needed you could perhaps have MacPorts' use its default

Reply:
I was looking over upstream documentation and they seem to only talk about
64-Intel and 64-bit ARM. Most of the source code is Java, so in theory it should
work on 32-bit Intel and PowerPC, but I couldn't find any documentation to
confirm or deny that. I'll keep digging.

Reply:
It's from apple, so I'm sure they didn't care about any platforms that weren't x86_64, or arm64, but it could possibly work...

Reply:
Yea, I haven't dug too deeply into the source code, but from what I was able to discern from the CI build instructions, it supports only x86_64 and arm64 as native binary platforms. I don't have any evidence to disprove that assumption yet.

Reply:
From the build log, it looks like an ARM-only binary is being created:

```
xinstall: /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/pkl-0.25.1/pkl-cli/build/executable/pkl-macos-amd64 -> /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/destroot/opt/local/bin/pkl
```

Is that expected behavior...?

Reply:
That's how I wrote the Portfile, yea. It attempts to build the binary for the architecture of the build machine. I believe GraalVM has the capability of cross-compiling for both Intel and ARM, but since ports are typically installed by compiling on the target machine, I figured only the active architecture was required.

Reply:
> That's how I wrote the Portfile, yea. It attempts to build the binary for the architecture of the build machine. I believe GraalVM has the capability of cross-compiling for both Intel and ARM, but since ports are typically installed by compiling on the target machine, I figured only the active architecture was required.

Bear in mind that the GitHub CI builders are all Intel. So if the arch is being properly set, an Intel binary should be created.

Haven't dug into upstream's build logic. But it may simply require passing in appropriate arg(s), to ensure the correct arch is built.

Reply:
> Bear in mind that the GitHub CI builders are all Intel. So if the arch is being properly set, an Intel binary should be created.

Are they? I see that the macOS 14 runner [reports an ARM architecture](https://github.com/macports/macports-ports/actions/runs/7834486473/job/21377807780?pr=22592#step:8:102)

```
DEBUG: OS darwin/23.2.0 (macOS 14.2.1) arch arm
```

In that same job, I see the `macExecutableAarch64` target [gets executed](https://github.com/macports/macports-ports/actions/runs/7834486473/job/21377807780?pr=22592#step:8:298):

```
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/pkl-0.25.1" && ./gradlew --info --stacktrace pkl-cli:macExecutableAarch64 -DreleaseBuild=true 
```

Looking at the macOS 13 runner, it appears to be Intel:

```
DEBUG: OS darwin/22.6.0 (macOS 13.6.3) arch i386
```

And builds the `macExecutableAmd64` target:

```
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/pkl-0.25.1" && ./gradlew --info --stacktrace pkl-cli:macExecutableAmd64 -DreleaseBuild=true
```

This is why I assumed the architectures were built correctly for the build machine.

Reply:
>> Bear in mind that the GitHub CI builders are all Intel. So if the arch is being properly set, an Intel binary should be created.
> 
> Are they? I see that the macOS 14 runner [reports an ARM architecture](https://github.com/macports/macports-ports/actions/runs/7834486473/job/21377807780?pr=22592#step:8:102)
> 
> ```
> DEBUG: OS darwin/23.2.0 (macOS 14.2.1) arch arm
> ```
> 
> In that same job, I see the `macExecutableAarch64` target [gets executed](https://github.com/macports/macports-ports/actions/runs/7834486473/job/21377807780?pr=22592#step:8:298):
> 
> ```
> DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/pkl-0.25.1" && ./gradlew --info --stacktrace pkl-cli:macExecutableAarch64 -DreleaseBuild=true 
> ```
> 
> Looking at the macOS 13 runner, it appears to be Intel:
> 
> ```
> DEBUG: OS darwin/22.6.0 (macOS 13.6.3) arch i386
> ```
> 
> And builds the `macExecutableAmd64` target:
> 
> ```
> DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_pkl/pkl/work/pkl-0.25.1" && ./gradlew --info --stacktrace pkl-cli:macExecutableAmd64 -DreleaseBuild=true
> ```
> This is why I assumed the architectures were built correctly for the build machine.

Ah yep, you're right... everything matches up.

So never mind, we're good!

Reply:
As for wrapping up the overall arch topic:

* The `supported_archs` value looks correct.
* For your architecture check, you'll want to base it on `configure.build_arch` (representing the desired target architecture), rather than `os.arch`:

```
if {${configure.build_arch} eq "arm64"} {
    # ARM build
} else {
    # Intel build
}
```

Let me know if that make sense.

Reply:
Yep! Makes sense to me.